### PR TITLE
feat(DTFS2-7298): hide cancel deal button when deal cancelled

### PR DESCRIPTION
**Pull request by BethThomas141**
## Introduction :pencil2:
- add `canDealStillBeCancelled` to check that the cancellation stage isn't COMPLETED or SCHEDULED (ie. check it hasn't already be cancelled. 
- Impact of this is that the 'Cancel deal' button will be hidden when the deal is already cancelled

## Resolution :heavy_check_mark:
List all changes made to the codebase.

## Miscellaneous :heavy_plus_sign:
List any additional fixes or improvements.

### CODE REVIEW

File: libs/common/src/constants/index.ts [hunk 1]
```diff
@@ -24,6 +24,5 @@ export * from './url-character-allowances';
 export * from './max-character-count';
 export * from './deal-submission-type';
 export * from './floating-point-rounding-decimal-places';
-export * from './cancellation-status';
```
Comment (BethThomas141) [inline libs/common/src/constants/index.ts:27]: I moved the deal cancellation status constants inside the `tfm` folder which is why these changes are showing up

File: trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ts [hunk 2]
```diff
@@ -16,14 +23,19 @@ export const canSubmissionTypeBeCancelled = (submissionType: DealSubmissionType)
  * Checks if deal cancellation is enabled for a deal and user type
  * @param submissionType - the deal submission type
  * @param user - the session user
+ * @param cancellationStatus - the current deal cancellation status if it exists
  * @returns true or false depending on the feature flag, submission type and user type
  */
-export const isDealCancellationEnabled = (submissionType: DealSubmissionType, user: TfmSessionUser): boolean => {
-  // TODO: DTFS2-7298: also check that the deal hasn't already been cancelled.
+export const canDealBeCancelled = (submissionType: DealSubmissionType, user: TfmSessionUser, cancellationStatus?: TfmDealCancellationStatus): boolean => {
```
Comment (BethThomas141) [inline trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ts:29]: I've renamed this as I think the new name is a tad clearer (makes it clear it's not just based on the feature flag being enabled)

File: trade-finance-manager-ui/server/controllers/case/index.js [hunk 3]
```diff
@@ -45,13 +44,11 @@ const getCaseDeal = async (req, res) => {
...
-  let hasDraftCancellation = false;
+  const dealCancellation = await api.getDealCancellation(dealId, userToken);
 
-  if (dealCancellationIsEnabled) {
-    const cancellation = await api.getDealCancellation(dealId, userToken);
-    hasDraftCancellation = !isEmpty(cancellation);
-  }
+  const dealCancellationPermitted = canDealBeCancelled(submissionType, user, dealCancellation.status);
+
+  const hasDraftCancellation = dealCancellationPermitted && dealCancellation?.status === TFM_DEAL_CANCELLATION_STATUS.DRAFT;
```
Comment (ttbarnes) [inline trade-finance-manager-ui/server/controllers/case/index.js:51]: Maybe could move into a helper directory, like `canDealBeCancelled`?

File: trade-finance-manager-ui/server/controllers/case/index.test.js [hunk 4]
```diff
@@ -99,12 +99,12 @@ describe('controllers - case', () => {
       it('should check whether deal cancellation is enabled', async () => {
         await caseController.getCaseDeal(req, res);
 
-        expect(isDealCancellationEnabled).toHaveBeenCalledTimes(1);
+        expect(canDealBeCancelled).toHaveBeenCalledTimes(1);
```
Comment (ttbarnes) [inline trade-finance-manager-ui/server/controllers/case/index.test.js:102]: Should we add `calledWith()` assertions?

File: trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ff-test.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+import { TfmSessionUser } from '../../types/tfm-session-user';
+
+const pimUser = { teams: [TEAM_IDS.PIM] } as TfmSessionUser;
+const nonPimUser = { teams: [TEAM_IDS.UNDERWRITERS] } as TfmSessionUser;
+
+const { AIN, MIN, MIA } = DEAL_SUBMISSION_TYPE;
+const { COMPLETED, SCHEDULED, DRAFT } = TFM_DEAL_CANCELLATION_STATUS;
+
+describe('canDealBeCancelled', () => {
+  describe('when `FF_TFM_FACILITY_END_DATE_ENABLED` is set to true', () => {
```
Comment (ttbarnes) [inline trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ff-test.ts:12]: We have a lot of inconsistent test descriptions for feature flag tests:

```ts
describe('when deal cancellation is enabled', () => {
describe('when deal cancellation is disabled', () => {
  it('does not call next when deal cancellation is enabled', () => {

describe('when `FF_TFM_FACILITY_END_DATE_ENABLED` is set to true', () => {
describe('when FF_TFM_DEAL_CANCELLATION_ENABLED is disabled', () => {
describe('when FF_TFM_DEAL_CANCELLATION_ENABLED is enabled', () => {

it("redirects to '/not-found' if the feature flag is not enabled", () => {
it('calls the next function if the feature flag is enabled', () => {
```

Not sure what to suggest here, we should align these in a different PR

File: trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ts [hunk 6]
```diff
@@ -16,14 +23,19 @@ export const canSubmissionTypeBeCancelled = (submissionType: DealSubmissionType)
  * Checks if deal cancellation is enabled for a deal and user type
  * @param submissionType - the deal submission type
  * @param user - the session user
+ * @param cancellationStatus - the current deal cancellation status if it exists
```
Comment (ttbarnes) [inline trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ts:26]:
```suggestion
 * @param cancellationStatus - optional current deal cancellation status
```

File: trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ts [hunk 7]
```diff
@@ -16,14 +23,19 @@ export const canSubmissionTypeBeCancelled = (submissionType: DealSubmissionType)
...
+export const canDealBeCancelled = (submissionType: DealSubmissionType, user: TfmSessionUser, cancellationStatus?: TfmDealCancellationStatus): boolean => {
+  const isDealCancellationFeatureFlagEnabled = isTfmDealCancellationFeatureFlagEnabled();
 
   const isUserAllowedToCancelDeal = userIsInTeam(user, [TEAM_IDS.PIM]);
-  const isDealCancellationFeatureFlagEnabled = isTfmDealCancellationFeatureFlagEnabled();
+
   const isAcceptableSubmissionType = canSubmissionTypeBeCancelled(submissionType);
 
-  return isUserAllowedToCancelDeal && isDealCancellationFeatureFlagEnabled && isAcceptableSubmissionType;
+  const isDealNotAlreadyCancelled = !(
```
Comment (ttbarnes) [inline trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ts:36]: Finding `isDealNotAlreadyCancelled` a little confusing especially in the final return, how about something like this?

```ts
const dealIsAlreadyCancelled = Object.values(TFM_DEAL_CANCELLATION_STATUS).includes(cancellationStatus);
```

Actually `Object.values` probably won't work here, maybe just remove the `!` and rename the existing const?

Then the return can do `!dealIsAlreadyCancelled`

**Review by ttbarnes — CHANGES_REQUESTED** [review]
Nice work - few minor comments

Can we add an E2E test?

File: libs/common/src/constants/index.ts (hunk 1 above)
Comment (MarRobSoftwire) [reply libs/common/src/constants/index.ts:27]: Good spot - thanks! Tbf the name should probably be tfm-cancellation-status too oops

File: libs/common/src/types/tfm/cancellation-status.ts [hunk 8]
```diff
@@ -0,0 +1,4 @@
+import { TFM_DEAL_CANCELLATION_STATUS } from '../../constants';
+import { ValuesOf } from '..';
+
+export type TfmDealCancellationStatus = ValuesOf<typeof TFM_DEAL_CANCELLATION_STATUS>;
```
Comment (MarRobSoftwire) [inline libs/common/src/types/tfm/cancellation-status.ts:4]: I probably should have done this too oops! Could you use this type in `TfmDealCancellationWithStatus` please :)

File: trade-finance-manager-ui/server/controllers/case/index.js [hunk 9]
```diff
@@ -45,13 +44,11 @@ const getCaseDeal = async (req, res) => {
...
   const { submissionType } = deal.dealSnapshot;
 
-  const dealCancellationIsEnabled = isDealCancellationEnabled(submissionType, user);
-  let hasDraftCancellation = false;
+  const dealCancellation = await api.getDealCancellation(dealId, userToken);
 
-  if (dealCancellationIsEnabled) {
-    const cancellation = await api.getDealCancellation(dealId, userToken);
-    hasDraftCancellation = !isEmpty(cancellation);
-  }
```
Comment (MarRobSoftwire) [inline trade-finance-manager-ui/server/controllers/case/index.js:54]: Checking the status rather than using `isEmpty` is soooo much nicer - maybe too much for this PR but there's a few other places I added this check that could now look for draft (and that would prevent the pages being accessed once the cancellations submitted)

File: trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ts (hunk 7 above)
Comment (MarRobSoftwire) [reply trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ts:36]: Agree `dealIsAlreadyCancelled` is much easier to read :) And since we're negating here anyway it's probably simpler to see whats going on anyway

File: trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.test.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+const nonPimUser = { teams: [TEAM_IDS.UNDERWRITERS] } as TfmSessionUser;
+
+const { AIN, MIN, MIA } = DEAL_SUBMISSION_TYPE;
+const { COMPLETED, SCHEDULED, DRAFT } = TFM_DEAL_CANCELLATION_STATUS;
+
+describe('canDealBeCancelled', () => {
+  describe('when `FF_TFM_FACILITY_END_DATE_ENABLED` is set to false', () => {
+    describe('when user is a PIM user', () => {
+      it.each([MIA, MIN, AIN])('should return false if the deal type is %s and the cancellation status is completed', (type) => {
+        const result = canDealBeCancelled(type, pimUser, COMPLETED);
```
Comment (MarRobSoftwire) [inline trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.test.ts:15]: could these `it.each`'s be `describe.each`?

```suggestion
      describe.each([MIA, MIN, AIN])('when the deal type is %s', (type) => {
        it('should return false if the cancellation status is completed', () => {
          const result = canDealBeCancelled(type, pimUser, COMPLETED);

          expect(result).toEqual(false);
        });

        it('should return false if the cancellation status is scheduled', (type) => {
          const result = canDealBeCancelled(type, pimUser, SCHEDULED);

          expect(result).toEqual(false);
        });

        // etc...
      });
```

File: trade-finance-manager-ui/server/controllers/case/index.js (hunk 3 above)
Comment (BethThomas141) [reply trade-finance-manager-ui/server/controllers/case/index.js:51]: This logic will only be used here, so keen to keep it as is if that's ok!

File: trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.test.ts (hunk 10 above)
Comment (MarRobSoftwire) [reply trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.test.ts:15]: actually is this just copied over & git being weird? feel free to ignore

File: trade-finance-manager-ui/server/controllers/case/index.test.js (hunk 4 above)
Comment (BethThomas141) [reply trade-finance-manager-ui/server/controllers/case/index.test.js:103]: Have added, thank you!

File: trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ff-test.ts (hunk 5 above)
Comment (BethThomas141) [reply trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ff-test.ts:12]: yes good spot - seems like there's no clear precedent here so maybe one for another PR

File: trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ts (hunk 6 above)
Comment (BethThomas141) [reply trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ts]: changed, thanks!

File: trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ts (hunk 7 above)
Comment (BethThomas141) [reply trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.ts]: Have changed! Thanks both

File: libs/common/src/constants/index.ts (hunk 1 above)
Comment (BethThomas141) [reply libs/common/src/constants/index.ts:27]: Have updated the name!

File: trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.test.ts (hunk 10 above)
Comment (BethThomas141) [reply trade-finance-manager-ui/server/controllers/helpers/can-deal-be-cancelled.helper.test.ts]: Good spot, that is a valid and better change - have updated the test files. Thanks!

File: libs/common/src/types/tfm/cancellation-status.ts (hunk 8 above)
Comment (BethThomas141) [reply libs/common/src/types/tfm/cancellation-status.ts]: Yes good idea! Have changed, thanks

File: trade-finance-manager-ui/server/controllers/case/index.js (hunk 9 above)
Comment (BethThomas141) [reply trade-finance-manager-ui/server/controllers/case/index.js:54]: Yes I noticed that! Let's do that as a separate PR

File: trade-finance-manager-ui/server/controllers/case/index.test.js [hunk 11]
```diff
@@ -147,7 +149,7 @@ describe('controllers - case', () => {
 
       describe('when deal cancellation is disabled', () => {
         it('should render the template with showDealCancelButton=false', async () => {
-          jest.mocked(isDealCancellationEnabled).mockReturnValueOnce(false);
+          jest.mocked(canDealBeCancelled).mockReturnValueOnce(false);
```
Comment (ttbarnes) [inline trade-finance-manager-ui/server/controllers/case/index.test.js:152]: Can we update the test descriptions please @BethThomas141 , they're currently:

- when deal cancellation is enabled
- when deal cancellation is disabled

(etc)

But the condition is now "is deal cancellation permitted"

Comment (ttbarnes) [inline trade-finance-manager-ui/server/controllers/case/index.test.js:152]: Also, unless i'm missing it, could we add a test for "deal cancellation is permitted, but the deal cancellation status is _not_ `TFM_DEAL_CANCELLATION_STATUS.DRAFT` - for the `hasDraftCancellation` page/nunjucks variable

The alternative would be to extract the `hasDraftCancellation` const into it's own helper function and then that could just be called in the test once

Comment (BethThomas141) [reply trade-finance-manager-ui/server/controllers/case/index.test.js:152]: Great spot, these tests needed refactoring after updating the logic in the index file (explained why I had a failing test😅). Have now rewritten to reflect the draft cancellation status check

Comment (BethThomas141) [reply trade-finance-manager-ui/server/controllers/case/index.test.js:152]: Great spot, have changed! Thanks!

**Review by ttbarnes — DISMISSED** [review]

**Review by ttbarnes — APPROVED** [review]

**Review by ttbarnes — APPROVED** [review]

File: trade-finance-manager-ui/server/controllers/helpers/deal-cancellation-enabled.helper.ts [hunk 12]
```diff
@@ -19,11 +26,29 @@ export const canSubmissionTypeBeCancelled = (submissionType: DealSubmissionType)
...
 
   return isUserAllowedToCancelDeal && isDealCancellationFeatureFlagEnabled && isAcceptableSubmissionType;
 };
+
+/**
+ * Checks if a deal can be cancelled; if it has already been cancelled then this returns false
+ * @param cancellationStatus - optional current deal cancellation status
+ * @returns true or false depending on if the deal has already been cancelled
+ */
+export const canDealStillBeCancelled = (cancellationStatus?: TfmDealCancellationStatus): boolean => {
```
Comment (ttbarnes) [inline trade-finance-manager-ui/server/controllers/helpers/deal-cancellation-enabled.helper.ts:41]: What's the reason for `canDealStillBeCancelled` please, how about `canDealBeCancelled`?

File: trade-finance-manager-ui/server/controllers/helpers/deal-cancellation-enabled.helper.ts [hunk 13]
```diff
@@ -19,11 +26,29 @@ export const canSubmissionTypeBeCancelled = (submissionType: DealSubmissionType)
...
+  const isDealAlreadyCancelled = cancellationStatus === TFM_DEAL_CANCELLATION_STATUS.COMPLETED || cancellationStatus === TFM_DEAL_CANCELLATION_STATUS.SCHEDULED;
+
+  return !isDealAlreadyCancelled;
```
Comment (ttbarnes) [inline trade-finance-manager-ui/server/controllers/helpers/deal-cancellation-enabled.helper.ts:44]: What do you think about this instead? More lines, but I think it's clearer

```ts
const isCompleted = cancellationStatus === TFM_DEAL_CANCELLATION_STATUS.COMPLETED;
const isScheduled = cancellationStatus === TFM_DEAL_CANCELLATION_STATUS.SCHEDULED;

const canBeCancelled = !isCompleted && !isScheduled;

return canBeCancelled;
```

File: trade-finance-manager-ui/server/controllers/helpers/deal-cancellation-enabled.helper.ts (hunk 12 above)
Comment (BethThomas141) [reply trade-finance-manager-ui/server/controllers/helpers/deal-cancellation-enabled.helper.ts]: Fair point, have changed! Thank you

**Review by ttbarnes — APPROVED** [review]
